llvm/lib/Object/SYCLBIN.cpp

```diff
@@ -0,0 +1,400 @@
+//===- SYCLBIN.cpp - SYCLBIN binary format support --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Object/SYCLBIN.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace llvm::object;
+
+namespace {
+
+template <typename T> void BinaryWriteInteger(raw_ostream &OS, T Val) {
+  static_assert(std::is_integral_v<T>);
```

Comment:
nit: we could consider using SFINAE instead of the assert, but maybe this way the error is more clear

Reply:
Since the name is explicit about it needing to be an integer, I like to think the error from a static assertion is clearer. SFINAE would've been nicer if we wanted it to select between implementations.

Reply:
It may be a bit of a minor optimization, but this function passes `Val` by value, which for an arbitrary `T` may be expensive. For integers it may be faster than passing by reference and should never be any more expensive, so that is the reason for the restriction.